File: back/src/main/java/com/back/domain/review/validation/ValidRatingValidator.java
package com.back.domain.review.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class ValidRatingValidator implements ConstraintValidator<ValidRating, Float> {
    @Override
    public boolean isValid(Float value, ConstraintValidatorContext context) {
        if (value == null) return true;

        boolean isInRange = 0f <= value && value <= 5f;
        boolean isHalf = value * 2f % 1f == 0;

        return isInRange && isHalf;
    }
}

File: back/src/main/java/com/back/domain/review/validation/ValidRatingValidator.kt
package com.back.domain.review.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class ValidRatingValidator : ConstraintValidator<ValidRating, Float> {

    override fun isValid(value: Float?, context: ConstraintValidatorContext?): Boolean {
        if (value == null) return true

        val isInRange = value in 0f..5f
        val isHalf = value * 2f % 1f == 0f

        return isInRange && isHalf
    }
}